### ISSUE TITLE
Reliability problems

### ISSUE THREAD

**Issue by davidfowl** [dotnet/project-system#2202]
Several times a day I open up a solution to an error list and IDE experience that looks like this:

![image](https://cloud.githubusercontent.com/assets/95136/26142480/90965fde-3a95-11e7-840d-0ff5ee8d64d1.png)

Restore completed and everything seems to have gone fine but the IDE never recovers. My package manager output window *does* have this:

```
Error occurred while restoring NuGet packages: The operation failed as details for project CodeGenerator could not be loaded.
Error occurred while restoring NuGet packages: The operation failed as details for project CodeGenerator could not be loaded.
Error occurred while restoring NuGet packages: The operation failed as details for project CodeGenerator could not be loaded.
Error occurred while restoring NuGet packages: The operation failed as details for project CodeGenerator could not be loaded.
Error occurred while restoring NuGet packages: The operation failed as details for project CodeGenerator could not be loaded.
Error occurred while restoring NuGet packages: The operation failed as details for project CodeGenerator could not be loaded.
Error occurred while restoring NuGet packages: The operation failed as details for project CodeGenerator could not be loaded.
Error occurred while restoring NuGet packages: The operation failed as details for project CodeGenerator could not be loaded.
Error occurred while restoring NuGet packages: The operation failed as details for project CodeGenerator could not be loaded.
Restoring NuGet packages...
Time Elapsed: 00:00:02.8433369
========== Finished ==========
```

But I was told by the nuget team that this is benign and that it could be ignored. When I try to compile in the IDE, everything is broken. When I close and re-open the solution it's still broken, when I build from the command line, everything works.

How do I diagnose these sorts of problems? Is there a switch or mode I can turn on that will give me more data about design time builds? xproj has an a flag you could set while the the IDE was still running that showed all of the design time information in a special output window. Is there anything like this in the new project system?

PS: Closing VS entirely and re-opening made it work.

**Comment by davkean** [comment]
First of all, what build are you running? We [regressed rename](https://github.com/dotnet/roslyn/issues/19389) in 15.3, so make sure you're not running into that.
Secondly, it's somewhat expected right now that you have hundreds of errors when you first open the project. The experience is terrible and we're fixing it for 15.3. Unlike legacy we don't block on the first design-time build (which makes open faster, but experience worse) - so Roslyn isn't told about references until the first design-time build. The errors **should** go away once the design-time build catches up.

If the errors stick around after the design-time builds have finished, or just during normal usage, such added/removing files, renaming  or moving, copying files around  - then this (hopefully) will be fixed by the [reliability changes](https://github.com/dotnet/project-system/pull/2178)  we're making these as we speak. If you've got a somewhat reliable repro, I'll grab and test it out on the new bits when I'm finished. We'll be added logging for this portion as part of that.

Design-Time builds themselves [already have logging](https://github.com/dotnet/project-system/blob/master/docs/design-time-builds.md#visual-studio-2017) - but I suspect it's not failed design-time build related because we push these errors/warning into the Error List.

**Comment by davidfowl** [comment]
> First of all, what build are you running? We regressed rename in 15.3, so make sure you're not running into that.

The official preview. 15.3:

![image](https://cloud.githubusercontent.com/assets/95136/26143725/f48d259a-3a9a-11e7-9247-dbc99a5936c8.png)

> Secondly, it's somewhat expected right now that you have hundreds of errors when you first open the project. The experience is terrible and we're fixing it for 15.3. Unlike legacy we don't block on the first design-time build (which makes open faster, but experience is worse) - so Roslyn isn't told about references until the first design-time build. The errors should go away once the design-time build catches up.

2 things, I hope we fix that experience because it's super annoying, especially for big projects. The lesser of 2 evils might be to avoid initializing the language service until the project is fully initialized. In my case, the errors never went away.

> If the errors stick around after the design-time builds have finished, or just during normal usage, such added/removing files, renaming or moving, copying files around - then this (hopefully) will be fixed by the reliability changes we're making these as we speak. If you've got a somewhat reliable repro, I'll grab and test it out on the new bits when I'm finished. We'll be added logging for this portion as part of that.

Sweet. I'll see if I can grab newer bits to test. The repro isn't 100% consistent but I've seen it happen 3 or 4 times tonight on different projects.

> Design-Time builds themselves already have logging - but I suspect it's not failed design-time build related because we push these errors/warning into the Error List.

Can you explain why the command line build worked then? I really just want to compare what data the language service is getting compared to csc because something is out of sync. I'll try those instructions.

**Comment by davkean** [comment]
Don't try the later bits until Monday's build. 

> 2 things, I hope we fix that experience because it's super annoying, especially for big projects. The lesser of 2 evils might be to avoid initializing the language service until the project is fully initialized.
That's effectively what we're doing. We're going to give an experience of Misc project (ie if you open a file without a project) + project context, so no squiggles, etc for types we don't know about, but you still get intellisense for types within the project. We'll very likely introduce a cache as well, similar to the legacy project system to avoid that first design-time build on open if you've ever opened the project before.

If you open the project and the errors never go away, I've never seen that bug. I would be interested in that bug.

I suspect the issues you are running isn't the design-time build, but the the communication _after_ the design-time build -> Roslyn. That's where most of the bugs are. We have races between the results we get from evaluation and results we get from design-time builds, they will be fixed by Monday's build.

**Comment by davkean** [comment]
@davidfowl We've resolved most of these sorts of issues in 15.3 - if you are still running into, can file a new bug and we'll start from there?

**Comment by davidfowl** [comment]
Yep
